core/distributed/matrix.cpp

```diff
@@ -252,14 +258,105 @@ void Matrix<ValueType, LocalIndexType, GlobalIndexType>::read_distributed(
     GKO_ASSERT_EQ(comm.size(), col_partition->get_num_parts());
     auto exec = this->get_executor();
     auto local_part = comm.rank();
+    auto use_host_buffer = mpi::requires_host_buffer(exec, comm);
 
     // set up LinOp sizes
-    auto num_parts = static_cast<size_type>(row_partition->get_num_parts());
     auto global_num_rows = row_partition->get_size();
     auto global_num_cols = col_partition->get_size();
     dim<2> global_dim{global_num_rows, global_num_cols};
     this->set_size(global_dim);
 
+    device_matrix_data<value_type, global_index_type> all_data{exec};
+    if (assembly_type == assembly::communicate) {
+        size_type num_entries = data.get_num_stored_elements();
+        size_type num_parts = comm.size();
+        array<comm_index_type> overlap_count{exec, num_parts};
+        array<global_index_type> overlap_positions{exec, num_entries};
+        array<global_index_type> original_positions{exec, num_entries};
+        overlap_count.fill(0);
+        auto tmp_part = make_temporary_clone(exec, row_partition);
+        exec->run(matrix::make_count_overlap_entries(
+            data, tmp_part.get(), local_part, overlap_count, overlap_positions,
+            original_positions));
+
+        overlap_count.set_executor(exec->get_master());
+        std::vector<comm_index_type> overlap_send_sizes(
+            overlap_count.get_data(), overlap_count.get_data() + num_parts);
+        std::vector<comm_index_type> overlap_send_offsets(num_parts + 1);
+        std::vector<comm_index_type> overlap_recv_sizes(num_parts);
+        std::vector<comm_index_type> overlap_recv_offsets(num_parts + 1);
+
+        std::partial_sum(overlap_send_sizes.begin(), overlap_send_sizes.end(),
+                         overlap_send_offsets.begin() + 1);
+        comm.all_to_all(exec, overlap_send_sizes.data(), 1,
+                        overlap_recv_sizes.data(), 1);
+        std::partial_sum(overlap_recv_sizes.begin(), overlap_recv_sizes.end(),
+                         overlap_recv_offsets.begin() + 1);
+        overlap_send_offsets[0] = 0;
+        overlap_recv_offsets[0] = 0;
+
+        size_type n_send = overlap_send_offsets.back();
+        size_type n_recv = overlap_recv_offsets.back();
+        array<global_index_type> overlap_send_row_idxs{exec, n_send};
+        array<global_index_type> overlap_send_col_idxs{exec, n_send};
+        array<value_type> overlap_send_values{exec, n_send};
+        array<global_index_type> overlap_recv_row_idxs{exec, n_recv};
+        array<global_index_type> overlap_recv_col_idxs{exec, n_recv};
+        array<value_type> overlap_recv_values{exec, n_recv};
```

Comment:
I think this would mean that we would have to add to add a `set_executor` to `device_matrix_data` for the `use_host_buffer` check, and having the separate arrays is consistent with the arrays for `local_row_idxs`, `local)col_idxs` and so on just below. I would prefer leaving these arrays.